File: partition-array-for-maximum-sum/partition-array-for-maximum-sum.cs
public class Solution {
    public int MaxSumAfterPartitioning(int[] arr, int k) {
        if(arr.Length==0) return 0;
        
        int[] dp = new int[arr.Length];
        
        for(int i=arr.Length-1;i>=0;i--)
        {
            int? currvalue = null;
            int? max = null;
            for(int j=i;j<i+k && j<arr.Length;j++)
            {
                currvalue = (currvalue == null ?  arr[j] : Math.Max(currvalue.Value,arr[j]));
                int newpartition = currvalue.Value *(j-i+1) + (j!=arr.Length-1 ? dp[j+1] : 0);
                max = (max==null ? newpartition : Math.Max(newpartition,max.Value));
            }
            dp[i] = max.Value;
        }
        return dp[0];
        
    }
}

File: src/com/madhurima/PartitionArrayForMaximumSum.java
//Time complexity: O(n*k)
//Space Complexity: O(n)
//Did the code run successfully in LeetCode = yes


package com.madhurima;

public class PartitionArrayForMaximumSum {
}

class PartitionArrayForMaximumSum1 {
    public int maxSumAfterPartitioning(int[] arr, int k) {
        if(arr == null || arr.length == 0){
            return 0;
        }

        int max = 0;
        int n = arr.length;
        int[] dp = new int[n];
        dp[0] = arr[0];

        for(int i = 1; i < n; i++){
            max = arr[i];
            for(int j = 1; j <= k && i - j + 1 >= 0; j++){ //window size
                max = Math.max(max, arr[i-j+1]);
                if(i-j >= 0){
                    dp[i] = Math.max(dp[i], max*j + dp[i-j]);
                }else{
                    dp[i] = Math.max(dp[i], max*j);
                }

            }
        }

        return dp[n-1];
    }
}
